### PR TITLE
refactor: read only txn request will use read state mechanism

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
Let read-only `TxnRequest` be handled like a normal `RangeRequest`
* what changes does this pull request make?
Add a new branch to the processing of `TxnRequest`
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
no

### CODE REVIEW

File: xline/src/server/kv_server.rs [hunk 1]
```diff
@@ -470,19 +467,34 @@ where
...
-            Self::update_header_revision(&mut res, revision);
-        }
+        let res = if txn_req.is_read_only() {
+            debug!("TxnRequest is read only");
+            let is_serializable = txn_req.is_serializable();
+            let token = get_token(request.metadata());
+            let wrapper = RequestWithToken::new_with_token(request.into_inner().into(), token);
+            let propose_id = self.generate_propose_id();
+            let cmd = Self::command_from_request_wrapper(propose_id, wrapper);
+            if is_serializable {
```
Comment (bsbds) [inline xline/src/server/kv_server.rs:478]: should this be `if !is_serializable`?

File: xline/src/server/kv_server.rs [hunk 2]
```diff
@@ -401,7 +393,12 @@ where
         if !is_serializable {
             self.wait_read_state(&cmd).await?;
         }
-        self.serializable_range(cmd.request())
+        let res = self.do_serializable(cmd.request())?;
+        if let Response::ResponseRange(response) = res {
+            Ok(tonic::Response::new(response))
+        } else {
+            panic!("Receive wrong response {res:?} for RangeRequest");
```
Comment (Phoenix500526) [inline xline/src/server/kv_server.rs:400]: Does `unreachable!` make more sense?

File: xline/src/server/kv_server.rs [hunk 3]
```diff
@@ -470,19 +467,34 @@ where
...
+                debug!("Get revision {:?} for TxnRequest", revision);
+                Self::update_header_revision(&mut res, revision);
+            }
+            res
+        };
         if let Response::ResponseTxn(response) = res {
             Ok(tonic::Response::new(response))
         } else {
-            panic!("Receive wrong response {res:?} for TxnRequest");
+            panic!("Receive wrong response {res:?} for TxnResponse");
```
Comment (Phoenix500526) [inline xline/src/server/kv_server.rs:497]: Same as above.

File: xline/tests/kv_test.rs [hunk 4]
```diff
@@ -249,3 +249,55 @@ async fn test_kv_delete() -> Result<(), Box<dyn Error>> {
 
     Ok(())
 }
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 10)]
```
Comment (Phoenix500526) [inline xline/tests/kv_test.rs:253]: Why should we set `worker_threads` as 10? Its default value is the number of cpus on the system, so I guess keeping it as default is better.

File: xline/src/server/kv_server.rs (hunk 1 above)
Comment (Phoenix500526) [reply xline/src/server/kv_server.rs]: How come? We should do `wait_read_state` when a client sends us a serializable range request.

Comment (themanforfree) [reply xline/src/server/kv_server.rs]: Fixed， it should be `if !is_serializable`, we only need to do wait_read_state when the request is not a serializable request

File: xline/src/server/kv_server.rs (hunk 2 above)
Comment (themanforfree) [reply xline/src/server/kv_server.rs]: Fixed

File: xline/src/server/kv_server.rs (hunk 3 above)
Comment (themanforfree) [reply xline/src/server/kv_server.rs]: Fixed

File: xline/tests/kv_test.rs (hunk 4 above)
Comment (themanforfree) [reply xline/tests/kv_test.rs]: To unify with other tests, now I have unified all the tests without set `worker_threads`

File: xline/src/server/kv_server.rs [hunk 5]
```diff
@@ -470,19 +467,34 @@ where
...
+            let token = get_token(request.metadata());
+            let wrapper = RequestWithToken::new_with_token(request.into_inner().into(), token);
+            let propose_id = self.generate_propose_id();
+            let cmd = Self::command_from_request_wrapper(propose_id, wrapper);
+            if !is_serializable {
+                self.wait_read_state(&cmd).await?;
+            }
+            self.do_serializable(cmd.request())?
+        } else {
+            let is_fast_path = false; // lock need revision of txn
```
Comment (bsbds) [inline xline/src/server/kv_server.rs:483]: If we return the revision in execution result, can this be changed to use fast path

Comment (themanforfree) [reply xline/src/server/kv_server.rs:483]: Yes, you can change this in #391

**Review by Phoenix500526 — COMMENTED** [review]
Let’s assume that there are two committed log entries, `PUT A 1` and `PUT B 1`,  in our state machine log. Now, user_1 proposes a TXN read request (READ A and B) at moment A, while user_2 proposes a put request ate moment B. The moment A is before moment B, and their interval is very short. After user_1 finishes his request, what will he read? (A, B) = (1, 1) or (A, B) = (1, 2)?  It depends on the execution request of `wait_read_state`. Does it align with our initial expectations for the TXN?

File: xline/src/server/kv_server.rs [hunk 6]
```diff
@@ -534,22 +533,37 @@ where
         debug!("Receive grpc request: {:?}", txn_req);
         Self::check_txn_request(
             txn_req,
-            self.kv_storage.compacted_revision(),
+            self.kv_storage.revision(),
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/kv_server.rs]: Fixed

**Pull request closed without merging**